lambda-runtime/src/runtime.rs

```
@@ -133,6 +153,95 @@ impl<S> Runtime<S> {
+        // They are reported to Lambda via /invocation/{id}/error and the worker
+        // continues. This only captures unrecoverable runtime failures like
+        // API client failures, runtime panics, etc.
+        let mut first_error: Option<BoxError> = None;
+        let mut remaining_workers = limit;
+        while let Some(result) = futures::StreamExt::next(&mut workers).await {
+            remaining_workers = remaining_workers.saturating_sub(1);
+            match result {
+                Ok(Ok(())) => {
+                    // `concurrent_worker_loop` runs indefinitely, so an Ok return indicates
+                    // an unexpected worker exit; we still decrement because the task is gone.
+                    warn!(
+                        remaining_workers,
+                        "Concurrent worker exited cleanly (unexpected - loop should run forever)"
+                    );
+                    if first_error.is_none() {
+                        first_error = Some(Box::new(io::Error::other(
+                            "all concurrent workers exited cleanly (unexpected - loop should run forever)",
+                        )));
+                    }
+                }
+                Ok(Err(err)) => {
+                    error!(error = %err, remaining_workers, "Concurrent worker exited with error");
+                    if first_error.is_none() {
+                        first_error = Some(err);
+                    }
+                }
+                Err(join_err) => {
+                    let err: BoxError = Box::new(join_err);
+                    error!(error = %err, remaining_workers, "Concurrent worker panicked");
+                    if first_error.is_none() {
+                        first_error = Some(err);
+                    }
+                }
+            }
+        }
+
+        match first_error {
+            Some(err) => Err(err),
+            None => Ok(()),
```

(Couldn't leave feedback on `fn run()`):

We should probably return an error or log an error message or something if `run()` is invoked with concurrency set, no? Currently it silently ignores it which is fairly confusing.